src/main/java/pl/edu/icm/pl/mxrdr/extension/workflow/XdsSupplyMissingValuesStep.java

```diff
@@ -0,0 +1,137 @@
+    }
+
+    private Optional<DatasetVersion> findVersion(WorkflowExecutionContext executionContext) {
+        WorkflowExecution execution = executionContext.getExecution();
+        DatasetVersion version = versionService.findByVersionNumber(execution.getDatasetId(),
+                execution.getMajorVersionNumber(), execution.getMinorVersionNumber());
+        return Optional.ofNullable(version);
+    }
+
+    // -------------------- INNER CLASSES --------------------
+
+    private static class Initializer {
+        private static final String UNDEFINED_REGEX = "(XXX|0[^.]?)";
+        private static final String COMMENT = " ! value supplied by " + STEP_ID + " step.";
+
+        static Set<String> initializeSourceFieldsSet() {
+            return Stream.of(MxrdrMetadataField.DATA_COLLECTION_DETECTOR_DISTANCE,
+                    MxrdrMetadataField.DATA_COLLECTION_OSCILLATION_STEP_SIZE,
+                    MxrdrMetadataField.DATA_COLLECTION_STARTING_ANGLE,
+                    MxrdrMetadataField.DATA_COLLECTION_WAVE_LENGTH,
+                    MxrdrMetadataField.DATA_COLLECTION_ORG_X,
+                    MxrdrMetadataField.DATA_COLLECTION_ORG_Y)
+                    .map(MxrdrMetadataField::getValue)
+                    .collect(Collectors.toSet());
+        }
+
+        static Map<String, PatternAndReplacement> initializePatternMap() {
+            Map<String, PatternAndReplacement> patterns = new HashMap<>();
+            patterns.put(MxrdrMetadataField.DATA_COLLECTION_DETECTOR_DISTANCE.getValue(),
+                    FileContentReplacer.PatternAndReplacement.of("^.*DETECTOR_DISTANCE=\\s*" + UNDEFINED_REGEX,
+                            "DETECTOR_DISTANCE= %s" + COMMENT));
+            patterns.put(MxrdrMetadataField.DATA_COLLECTION_OSCILLATION_STEP_SIZE.getValue(),
+                    FileContentReplacer.PatternAndReplacement.of("^.*OSCILLATION_RANGE=\\s*" + UNDEFINED_REGEX,
+                            "OSCILLATION_RANGE= %s" + COMMENT));
+            patterns.put(MxrdrMetadataField.DATA_COLLECTION_STARTING_ANGLE.getValue(),
+                    FileContentReplacer.PatternAndReplacement.of("^.*STARTING_ANGLE=\\s*" + UNDEFINED_REGEX,
+                            "STARTING_ANGLE= %s" + COMMENT));
+            patterns.put(MxrdrMetadataField.DATA_COLLECTION_WAVE_LENGTH.getValue(),
+                    FileContentReplacer.PatternAndReplacement.of("^.*X-RAY_WAVELENGTH=\\s*" + UNDEFINED_REGEX,
+                            "X-RAY_WAVELENGTH= %s" + COMMENT));
+            patterns.put(MxrdrMetadataField.DATA_COLLECTION_ORG_X.getValue(),
+                    FileContentReplacer.PatternAndReplacement.of("^.*ORGX=\\s*" + UNDEFINED_REGEX,
+                            "ORGX= %s"));
+            patterns.put(MxrdrMetadataField.DATA_COLLECTION_ORG_Y.getValue(),
+                    FileContentReplacer.PatternAndReplacement.of("ORGY=\\s*" + UNDEFINED_REGEX,
+                            "ORGY= %s" + COMMENT));
```

Comment:
We can do with much simpler map here, only mapping `MxrdrMetadataField` to the name of the field in the file, i.e. `DATA_COLLECTION_DETECTOR_DISTANCE` => `DETECTOR_DISTANCE`. The actual pattern seems repeated with only this small difference of the field name. Only last one is missing `^.*` - is it intentional?

Reply:
>Only last one is missing ^.* - is it intentional?

Yes, because in one case we have two values in one line.

Reply:
However it could be done also in this other way.